File: packages/shared/lib/src/models/payment.dart
import '../enums/payment_status.dart';

class Payment {
  final String id;
  final String? invoiceId;
  final String? shipmentId;
  final double amount;
  final String paymentMethod;
  final PaymentStatus status;
  final String? transactionId;
  final String? collectedBy;
  final DateTime createdAt;

  Payment({
    required this.id,
    this.invoiceId,
    this.shipmentId,
    required this.amount,
    required this.paymentMethod,
    required this.status,
    this.transactionId,
    this.collectedBy,
    required this.createdAt,
  });

  factory Payment.fromJson(Map<String, dynamic> json) {
    return Payment(
      id: json['id'] as String,
      invoiceId: json['invoice_id'] as String?,
      shipmentId: json['shipment_id'] as String?,
      amount: (json['amount'] as num).toDouble(),
      paymentMethod: json['payment_method'] as String,
      status: PaymentStatus.fromApi(json['status'] as String),
      transactionId: json['transaction_id'] as String?,
      collectedBy: json['collected_by'] as String?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'invoice_id': invoiceId,
      'shipment_id': shipmentId,
      'amount': amount,
      'payment_method': paymentMethod,
      'status': status.apiValue,
      'transaction_id': transactionId,
      'collected_by': collectedBy,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

File: backend/api/app/models/payment.py
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, Float, ForeignKey, String, func
from app.database import GUID
from sqlalchemy.orm import Mapped, mapped_column

from app.database import Base

import enum


class PaymentMethod(str, enum.Enum):
    cash = "cash"
    credit_card = "credit_card"
    bank_transfer = "bank_transfer"
    online = "online"


class PaymentStatus(str, enum.Enum):
    pending = "pending"
    completed = "completed"
    failed = "failed"
    refunded = "refunded"


class Payment(Base):
    __tablename__ = "payments"

    id: Mapped[uuid.UUID] = mapped_column(GUID(), primary_key=True, default=uuid.uuid4)
    invoice_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("invoices.id"), nullable=True)
    shipment_id: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("shipments.id"), nullable=True)
    amount: Mapped[float] = mapped_column(Float, nullable=False)
    payment_method: Mapped[PaymentMethod] = mapped_column(Enum(PaymentMethod), nullable=False)
    status: Mapped[PaymentStatus] = mapped_column(Enum(PaymentStatus), default=PaymentStatus.pending)
    transaction_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    collected_by: Mapped[uuid.UUID | None] = mapped_column(GUID(), ForeignKey("users.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
